apps/main/src/dex/components/PagePool/Deposit/components/FieldsDeposit.tsx

```diff
@@ -1,13 +1,57 @@
+import { BigNumber } from 'bignumber.js'
+import { zip } from 'lodash'
 import cloneDeep from 'lodash/cloneDeep'
 import { useMemo } from 'react'
 import FieldToken from '@/dex/components/PagePool/components/FieldToken'
 import type { FormValues, LoadMaxAmount } from '@/dex/components/PagePool/Deposit/types'
 import { FieldsWrapper } from '@/dex/components/PagePool/styles'
 import type { TransferProps } from '@/dex/components/PagePool/types'
 import useStore from '@/dex/store/useStore'
+import type { CurrencyReserves } from '@/dex/types/main.types'
+import { getChainPoolIdActiveKey } from '@/dex/utils'
 import Checkbox from '@ui/Checkbox'
 import { formatNumber } from '@ui/utils'
 import { t } from '@ui-kit/lib/i18n'
+import { Amount } from '../../utils'
+
+/**
+ * Format the precision of the balanced value based on the USD price.
+ * The amount of decimals is determined by the USD price, so it's accurate to around $1.
+ */
+function formatPrecision(balancedValue: BigNumber, usdPrice: number) {
+  const decimals = Math.ceil(Math.log10(usdPrice))
+  return balancedValue.toFormat(decimals, BigNumber.ROUND_HALF_DOWN, { groupSeparator: '', decimalSeparator: '.' })
+}
+
+/**
+ * Calculate new balanced form values based on the changed index and value, keeping the ratios of the other tokens.
+ * This function is used to update the amounts in the form when a user changes the value of one of the inputs.
+ */
+function calculateBalancedValues(
+  [value, changedIndex]: [string, number],
+  oldAmounts: FormValues['amounts'],
+  tokenAddresses: string[],
+  { tokens, totalUsd }: CurrencyReserves,
+): Amount[] {
+  const reserves = Object.fromEntries(
+    tokens.map((t) => [t.tokenAddress, { usdPrice: t.usdRate, reserveRatio: t.balanceUsd / Number(totalUsd) }]),
+  )
+  const { reserveRatio: changedRatio, usdPrice: changedUsdPrice } = reserves[tokenAddresses[changedIndex]] ?? {}
+  return zip(oldAmounts, tokenAddresses).map((tuple, index) => {
+    const [amount, tokenAddress] = tuple as [Amount, string]
+    if (changedIndex === index) {
+      return { ...amount, value }
+    }
+    const { usdPrice, reserveRatio } = reserves[tokenAddress] ?? {}
+    if (usdPrice && reserveRatio && changedRatio) {
+      const valueUsd = BigNumber(value).times(changedUsdPrice)
```

Comment:
Is `BigNumber` really necessary or can we use javascript's native `bigint` here?

Reply:
These are floats

Reply:
tbh the calculation is probably precise enough with normal JS ints